### DISCUSSION TITLE
How do I rewrite global variables before loading a BPF object?

### DISCUSSION THREAD

**Discussion by Ephemeraler** [cilium/ebpf#795] [Q&A] [answered]
Before that, I use libbpf and C to write bpf controller. In *.bpf.c, I declare a global variable (filter_pid). and I set value (skel->bss->filter_pid = 1234) in *.c. Now, what should I do in cilium/ebpf. when I executed go generate, I might not find this variable  In bpf_**.go. 3Q.

**Comment by ti-mo** [comment, marked as answer]
@Ephemeraler ebpf-go implements this using `CollectionSpec.RewriteConstants()`.

Given the following [global variable in C](https://github.com/cilium/ebpf/blob/aaa9af24fdd8e397be823526928782cc93d59188/testdata/loader.c#L115):
```c
volatile const uint32_t arg;
```

The [following Go code](https://github.com/cilium/ebpf/blob/aaa9af24fdd8e397be823526928782cc93d59188/elf_reader_test.go#L179) takes care of rewriting that constant to hold the value `1` at runtime:
```go
err := spec.RewriteConstants(map[string]interface{}{
	"arg":  uint32(1),
})
```

**Reply by Ephemeraler** [reply]
Is there a manual? I really don't know how to do that. I add that code, and build. This is no error, but the value isn't be setted. Using examples/kprobe, I use `LoadCollectionSpec`  to load elf and`err := spec.RewriteConstants(map[string]interface{}{
	"arg":  uint32(1),
})`, involving bpf_printk(arg) in kprobe.c, but the arg value was always 0.

**Reply by ti-mo** [reply]
If there's no error calling `RewriteConstants`, ebpf-go found the variable in the ELF and successfully managed to replace it. It's a rather straightforward operation under the hood. That's kind of where the lib's scope and responsibility ends.

Hard to tell what's going on without the full C code, but `bpf_printk()` can't be used the way you describe. It takes a string pointer as its first argument. If it's a null pointer at compile time, hard to tell what the compiler will do.

I'd recommend looking at the bpf selftests (https://elixir.bootlin.com/linux/latest/source/tools/testing/selftests/bpf/progs) (the `_kern*.c` files) for guidance on how to use most BPF helpers. Hope that helps.

**Reply by Ephemeraler** [reply]
Oh, I use example/kprobe, the elf and bpf_*.go is compiled by go generate, and I write **RewriteConstants** in main.go, is it right?
```
// kprobe.c
#include "common.h"
char __license[] SEC("license") = "Dual MIT/GPL";

struct bpf_map_def SEC("maps") kprobe_map = {
    .type        = BPF_MAP_TYPE_ARRAY,
    .key_size    = sizeof(u32),
    .value_size  = sizeof(u64),
    .max_entries = 1,
};

volatile const int uid1 = 1009;

SEC("kprobe/sys_execve")
int kprobe_execve() {
    u32 key     = 0;
    u64 initval = 1, *valp;
    valp = bpf_map_lookup_elem(&kprobe_map, &key);
    if (!valp) {
        bpf_map_update_elem(&kprobe_map, &key, &initval, BPF_ANY);
        return 0;
    }
    __sync_fetch_and_add(valp, 1);
    bpf_printk("uid = %d", uid1);
    return 0;
}```


```
```
// main.go
package main

import (
    "log"
    "fmt"
    "time"
    "github.com/cilium/ebpf/link"
    "github.com/cilium/ebpf/rlimit"
    "github.com/cilium/ebpf"
)

// $BPF_CLANG and $BPF_CFLAGS are set by the Makefile.
//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -cc clang-10 -cflags "-O2 -g -Wall -Werror" --target=amd64 bpf kprobe.c -- -I ../headers


const mapKey uint32 = 0

func main() {

    // Name of the kernel function to trace.
    fn := "sys_execve"

    // Allow the current process to lock memory for eBPF resources.
    if err := rlimit.RemoveMemlock(); err != nil {
        log.Fatal(err)
    }

    have, err := ebpf.LoadCollectionSpec("bpf_bpfel_x86.o")
    if have.Types != nil {

        err := have.RewriteConstants(map[string]interface{}{
                "uid1":  int(100),
        })

        if err != nil {

            fmt.Println("Can't rewrite constant")
        }

    }
    // Load pre-compiled programs and maps into the kernel.
    objs := bpfObjects{}
    if err := loadBpfObjects(&objs, nil); err != nil {
        log.Fatalf("loading objects: %v", err)
    }
    defer objs.Close()
    kp, err := link.Kprobe(fn, objs.KprobeExecve, nil)
    if err != nil {
        log.Fatalf("opening kprobe: %s", err)
    }
    defer kp.Close()

    // Read loop reporting the total amount of times the kernel
    // function was entered, once per second.
    ticker := time.NewTicker(1 * time.Second)
    defer ticker.Stop()

    log.Println("Waiting for events..")

    for range ticker.C {
        var value uint64
        if err := objs.KprobeMap.Lookup(mapKey, &value); err != nil {
            log.Fatalf("reading map: %v", err)
        }
        log.Printf("%s called %d times\n", fn, value)
    }
}

**Reply by ti-mo** [reply]
Please fix your formatting.

**Reply by Ephemeraler** [reply]
have fixed

**Reply by Ephemeraler** [reply]
Maybe, I have some ideas. when calling LoadCollectionSpec and RewriteConstants, they  do not change file(bpf_bpfel_x86.o). So when I load bpf program, the variable(uid) not change. But, emmm, what should i do. rewrite elf file?

**Reply by Ephemeraler** [reply]
Thanks very much, I understand. I have solved.

**Reply by ti-mo** [reply]
This does not rewrite the ELF, the point is to inject values at runtime after loading the ELF. Glad you got it sorted.

**Reply by cougarwww** [reply]
> Thanks very much, I understand. I have solved.

How did you solved? i have the same error.

---
I have solved too.  we can not use bpfObjects after have.RewriteConstants , but use have.LoadAndAssign to get objs

**Reply by g0dA** [reply]
Yes, i have solved too, tks!

**Reply by devendra-dhakad** [reply]
@Ephemeraler @g0dA @cougarwww  Can you please share the code here?

**Reply by thediveo** [reply]
@g0dA can you please be so kind as to share the fixed code here? Thank you!

**Reply by thediveo** [reply]
For the benefit of others being stuck with the same problem here's how I solved it; it's a shame that the posters claiming to have a solution never shared their solutions in a clear manner, yet are expecting to get help in the future nevertheless. Anyway, here we Go, assuming that you want to use `bpf2go`:

The `ebpf` module examples usually use `loadBpfObjects(obj, opts)` in roughly this manner:

```go
objs := bpfObjects{}
err := loadBpfObjects(&objs, nil)
// ...error handling...
```

Luckily, `loadBpfObjects(...)` is a very simple wrapper around getting a `spec` by calling the autogenerated `loadBpf()` and then doing `spec.LoadAndAssign(...)`. Now, rewriting constants needs to be done _after_ loading the `spec` and _before_ `LoadAndAssign`:

```go
spec, err := loadBpf()
// ...error handling...
err = spec.RewriteConstants(map[string]interface{}{
  "foobar": uint32(42),
})
// ...error handling...
objs := bpfObjects{}
err = spec.LoadAndAssign(&objs, nil)
// ...error handling...
```

If something is unclear in my example, please ask!

**Reply by hossein-bakhtiari-revolut** [reply]
is it possible to set the elements of an array in bss using this technique?

**Reply by ti-mo** [reply]
@hossein-bakhtiari-revolut To be safe, always make section names part of the variable declaration:

`static volatile const uint32_t some_variable __section(".rodata.foo");`

Mind, `.rodata` is considered a prefix, you can use any name you'd like. Variables declared `const` should never end up in .bss, but the same advice applies for non-const globals. If you declare `static volatile`s, explicitly put them into `.data` to avoid them ending up in .bss if you leave them uninitialized.

**Reply by g0dA** [reply]
@thediveo sry, i just saw this email reminder

the code example
```
// in *.c
volatile const uint32_t arg;


// in *.go
err := spec.RewriteConstants(map[string]interface{}{
    "arg":  uint32(1),
})
```

but this variable will in `.rodata` section, this need `BPF_F_RDONLY`, was patched from [bpf: add program side {rd, wr}only support for maps](https://lore.kernel.org/bpf/20190409210910.32048-4-daniel@iogearbox.net/#r)

so, this feature also need your kernel version support

**Reply by thediveo** [reply]
can you please elaborate? my example works for me.

**Reply by g0dA** [reply]
useful is good, i just say that kernel version have effect, if your kernel version is low, you may get an error like:
```
map create: read- and write-only maps not supported (requires >= v5.2)
```

**Reply by thediveo** [reply]
thanks for clarifying!
